app/app/src/main/java/de/stustapay/stustapay/repository/InfallibleRepository.kt

```diff
@@ -0,0 +1,101 @@
+) {
+    private val scope: CoroutineScope =
+        CoroutineScope(Dispatchers.Default + CoroutineName("infallible"))
+    private lateinit var runner: Job
+
+    fun launch() {
+        runner = scope.launch {
+            dataSource.requests.collect {
+                var requests = it
+
+                var retry = true
+                var attempts = 3
+                while (retry && attempts > 0) {
+                    retry = false
+
+                    for ((id, request) in it) {
+                        when (process(request)) {
+                            InfallibleResult.Ok -> {
+                                dataSource.remove(id)
+                                requests = requests.filter { it.key != id }
+                            }
+
+                            InfallibleResult.Retry -> retry = true
+                        }
+                    }
+
+                    attempts -= 1
+                    delay(1000)
+                }
+            }
+        }
+    }
+
+    suspend fun bookTopUp(newTopUp: NewTopUp) {
+        dataSource.push(
+            newTopUp.uuid, InfallibleApiRequest(InfallibleApiRequestKind.TopUp(newTopUp))
+        )
+    }
+
+    suspend fun bookTicketSale(newTicketSale: NewTicketSale) {
+        dataSource.push(
+            newTicketSale.uuid,
+            InfallibleApiRequest(InfallibleApiRequestKind.TicketSale(newTicketSale))
+        )
+    }
+
+    private suspend fun process(request: InfallibleApiRequest): InfallibleResult {
+        val kind = request.kind
+        return when (kind) {
+            is InfallibleApiRequestKind.TopUp -> {
+                when (topUpRepository.bookTopUp(kind.content)) {
+                    is Response.OK -> InfallibleResult.Ok
+                    is Response.Error.Access -> InfallibleResult.Retry
+                    is Response.Error.NotFound -> InfallibleResult.Retry
+                    is Response.Error.Request -> InfallibleResult.Retry
+                    is Response.Error.Server -> InfallibleResult.Retry
+                    is Response.Error.Service.Generic -> InfallibleResult.Retry
+                    is Response.Error.Service.NotEnoughFunds -> InfallibleResult.Retry
+                }
+            }
+
+            is InfallibleApiRequestKind.TicketSale -> {
+                when (ticketRepository.bookTicketSale(kind.content)) {
+                    is Response.OK -> InfallibleResult.Ok
+                    is Response.Error.Access -> InfallibleResult.Retry
+                    is Response.Error.NotFound -> InfallibleResult.Retry
+                    is Response.Error.Request -> InfallibleResult.Retry
+                    is Response.Error.Server -> InfallibleResult.Retry
+                    is Response.Error.Service.Generic -> InfallibleResult.Retry
+                    is Response.Error.Service.NotEnoughFunds -> InfallibleResult.Retry
+                }
+            }
+        }
+    }
+}
+
+enum class InfallibleResult {
+    Ok, Retry
```

Comment:
it can still really fail, not just retry.

Reply:
Nope. The point of infallible requests is that they aren't allowed to fail. If a customer has already paid for a topup or a ticket, then we absolutely must book that transaction.